nmmo/task/game_state.py

```diff
@@ -0,0 +1,180 @@
+from __future__ import annotations
+from typing import Dict, List, Tuple, MutableMapping
+from dataclasses import dataclass
+from copy import deepcopy
+
+import numpy as np
+
+from nmmo.core.config import Config
+from nmmo.core.realm import Realm
+from nmmo.core.observation import Observation
+from nmmo.task.group import Group
+
+from nmmo.entity.entity import EntityState
+from nmmo.lib.event_log import EventState, ATTACK_COL_MAP, ITEM_COL_MAP, LEVEL_COL_MAP
+from nmmo.lib.log import EventCode
+from nmmo.systems.item import ItemState
+from nmmo.core.tile import TileState
+
+EntityAttr = EntityState.State.attr_name_to_col
+EventAttr = EventState.State.attr_name_to_col
+ItemAttr = ItemState.State.attr_name_to_col
+TileAttr = TileState.State.attr_name_to_col
+EventAttr.update(ITEM_COL_MAP)
+EventAttr.update(ATTACK_COL_MAP)
+EventAttr.update(LEVEL_COL_MAP)
+
+@dataclass(frozen=True) # make gs read-only, except cache_result
+class GameState:
+  current_tick: int
+  config: Config
+  spawn_pos: Dict[int, Tuple[int, int]] # ent_id: (row, col) of all spawned agents
+
+  alive_agents: List[int] # of alive agents' ent_id (for convenience)
+  env_obs: Dict[int, Observation] # env passes the obs of only alive agents
+
+  entity_data: np.ndarray # a copied, whole Entity ds table
+  item_data: np.ndarray # a copied, whole Item ds table
+  event_data: np.ndarray # a copied, whole Event log table
+
+  cache_result: MutableMapping # cache for general memoization
+  # add helper functions below
+  def entity_or_none(self, ent_id):
+    flt_ent = self.entity_data[:, EntityAttr['id']] == ent_id
+    if np.any(flt_ent):
+      return EntityAttr.parse_array(self.entity_data[flt_ent][0])
+
+    return None
+
+  def where_in_id(self, data_type, subject: List[int]):
+    if data_type == 'entity':
+      flt_idx = np.in1d(self.entity_data[:, EntityAttr['id']], subject)
+      return self.entity_data[flt_idx]
+
+    if data_type == 'item':
+      flt_idx = np.in1d(self.item_data[:, ItemAttr['owner_id']], subject)
+      return self.item_data[flt_idx]
+
+    if data_type == 'event':
+      flt_idx = np.in1d(self.event_data[:, EventAttr['ent_id']], subject)
+      return self.event_data[flt_idx]
+
+    return None
+
+  def get_subject_view(self, subject: Group):
+    return GroupView(self, subject)
+
+class ArrayView:
```

Comment:
Could we subclass instead of passing a "name"?

Reply:
Sub-classed into TileView, EventCodeView, EventView and EntityView but kept 'name' for generating caching keys and to avoid having a separate class for each Event type.